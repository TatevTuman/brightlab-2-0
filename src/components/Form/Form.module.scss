.form {

  &-item {
    display: flex;
    align-items: center;
    margin-bottom: 1rem;
  }

  &-submit {

  }
}
