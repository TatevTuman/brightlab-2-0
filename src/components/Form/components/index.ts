export { default as FormItem } from './FormItem/FormItem'
export { default as FormControl } from './FormControl/FormControl'
export { default as FormSubmit } from './FormSubmit/FormSubmit'
