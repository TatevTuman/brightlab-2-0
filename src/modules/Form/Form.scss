.form {
  margin-bottom: 0;

  &-item {
    display: flex;
    align-items: center;
    margin-bottom: 1rem;

    &:last-child {
      margin-bottom: 0 !important;
    }
  }

  &-control {
    width: 100%;

    &[data-no-width='true'] {
      width: unset;
    }
  }

  &-submit {

  }
}
