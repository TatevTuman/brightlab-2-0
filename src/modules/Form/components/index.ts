export { default as FormItem } from './FormItem/FormItem'
export { default as FormControl } from './FormControl'
export { default as FormSubmit } from './FormSubmit'

export type { FormItemProps } from './FormItem/FormItem'
export type { FormControlProps } from './FormControl'
export type { FormSubmitProps } from './FormSubmit'
