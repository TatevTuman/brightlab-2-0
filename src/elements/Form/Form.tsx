import React, { memo, forwardRef, ReactElement } from 'react'
import { useForm } from 'react-hook-form'
import { UnpackNestedValue, FieldErrors } from 'react-hook-form/dist/types/form'
import './Form.scss'

interface FormProps {
  defaultValues?: Record<string, any>
  children: JSX.Element | JSX.Element[]
  onSubmit: (data: UnpackNestedValue<Record<string, any>>) => void
  onError?: (errors: FieldErrors<Record<string, any>>) => void
}

const Form: React.ForwardRefExoticComponent<FormProps> = forwardRef<HTMLFormElement, FormProps>((props, ref) => {
  const { defaultValues, onSubmit, onError, children } = props
  const { register, handleSubmit, errors } = useForm({ defaultValues })

  return (
    <form ref={ref} onSubmit={handleSubmit(onSubmit, onError)}>
      {React.Children.map(children, (child: ReactElement) => {
        return child.props.name
          ? React.createElement(child.type, {
              ...{
                ...child.props,
                register: register(child.props.validation),
                error: errors[child.props.name],
                key: child.props.name
              }
            })
          : child
      })}
    </form>
  )
})

export default memo(Form)
