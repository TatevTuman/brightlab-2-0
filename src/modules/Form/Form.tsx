import React from 'react'
import {
  useForm,
  SubmitHandler,
  SubmitErrorHandler,
  UnpackNestedValue,
  DeepPartial,
  UseFormMethods,
  FieldName,
  FormProvider
} from 'react-hook-form'

import { Children } from '@types'
import { genericMemo } from '@hocs'
import './Form.scss'

export interface FormProps<F> {
  id?: string
  className?: string
  defaultValues?: UnpackNestedValue<DeepPartial<F>>
  children: (props: UseFormMethods<F>) => Children
  onSubmit: (form: UnpackNestedValue<F>, event?: React.BaseSyntheticEvent) => void
  onError: SubmitErrorHandler<F>
}

/*
  Form component that uses react-hook-form lib to manage form state without React state.

  Loading state is implemented via FormSubmit component that controls loading field like input in the form
  You can add here any logic you want to improve UX
*/

const Form = <F,>(props: FormProps<F>) => {
  const { id, className, defaultValues, onSubmit, onError, children } = props
  const useFormMethods = useForm<F>({ defaultValues })

  /* FormSubmit component controls loading field, so we can manage button state with react-hook-form */
  const setFormLoading = () =>
    setTimeout(
      () =>
        useFormMethods.setValue('loading' as FieldName<F>, false, {
          shouldValidate: false,
          shouldDirty: true
        }),
      500
    )

  /* Custom form submit handler */
  const handleFormSubmit: SubmitHandler<F> = async (form, event) => {
    await onSubmit(form, event)
    /* Set FormSubmit loading to false */
    setFormLoading()
  }

  /* Custom form error handler */
  const handleFormError: SubmitErrorHandler<F> = async (errors, event) => {
    await onError(errors, event)
    /* Set FormSubmit loading to false */
    setFormLoading()
  }

  return (
    <FormProvider {...useFormMethods}>
      <form
        id={id}
        className={'form ' + className}
        onSubmit={useFormMethods.handleSubmit<F>(handleFormSubmit, handleFormError)}
        noValidate
        data-testid={'form'}
      >
        {children(useFormMethods)}
      </form>
    </FormProvider>
  )
}

export default genericMemo(Form)
