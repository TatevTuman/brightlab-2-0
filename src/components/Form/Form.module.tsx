import { LoadableComponent } from '@loadable/component'
import { withLoadableFallback, WithLoadableFallbackOptions } from '@hocs'
import { FormControl, FormItem, FormSubmit } from './components'

import type { FormProps } from '@components'
import type { FormModule } from './Form'
type LoadableForm<F> = LoadableComponent<FormProps<F>> & FormModule

/*
  Form module
  TODO Research. Dot notation doesn't work with hocs

  Callback needs for generic form logic
  const SignInForm = Form<SignInFormType>()
*/

const Form = <F,>(withLoadableFallbackOptions?: WithLoadableFallbackOptions): LoadableForm<F> => {
  const FormModule = withLoadableFallback<FormProps<F>>(
    import('./Form'),
    withLoadableFallbackOptions || {}
  ) as LoadableForm<F>

  /* Adding component for module */
  FormModule.Control = FormControl
  FormModule.Item = FormItem
  FormModule.Submit = FormSubmit

  return FormModule
}

export default Form
