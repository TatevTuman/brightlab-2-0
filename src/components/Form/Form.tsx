import React, { memo, forwardRef, RefObject } from 'react'
import {
  useForm,
  SubmitHandler,
  SubmitErrorHandler,
  UnpackNestedValue,
  DeepPartial,
  UseFormMethods,
  FieldName,
  FormProvider
} from 'react-hook-form'
import { Children } from '@types'
import { FormItem, FormSubmit } from './components'
import styles from './Form.module.scss'

export interface FormProps<F> {
  defaultValues?: UnpackNestedValue<DeepPartial<F>>
  children: (props: UseFormMethods<F>) => Children
  onSubmit: SubmitHandler<F>
  onError: SubmitErrorHandler<F>
  ref?: RefObject<HTMLFormElement>
}

/*
  Form component that uses react-hook-form lib to manage form state without React state.
  Forward ref is used in awkward way because of Typescript genetic argument to pass it in useForm hook.
  TODO Research

  Loading state is implemented via FormSubmit component that controls loading field like input in the form
  You can add here any logic you want to improve UX
*/

const Form = <F,>(props: FormProps<F>) => {
  const ForwardRefWrapper = memo(
    forwardRef<HTMLFormElement, FormProps<F>>((props, ref) => {
      const { defaultValues, onSubmit, onError, children } = props
      const useFormMethods = useForm<F>({ defaultValues })

      /* FormSubmit component controls loading field, so we can manage button state with react-hook-form */
      const setFormLoading = () =>
        setTimeout(
          () =>
            useFormMethods.setValue('loading' as FieldName<F>, false, {
              shouldValidate: false,
              shouldDirty: true
            }),
          500
        )

      /* Custom form submit handler */
      const handleFormSubmit: SubmitHandler<F> = async (form, event) => {
        await onSubmit(form, event)
        /* Set FormSubmit loading to false */
        setFormLoading()
      }

      /* Custom form error handler */
      const handleFormError: SubmitErrorHandler<F> = async (errors, event) => {
        await onError(errors, event)
        /* Set FormSubmit loading to false */
        setFormLoading()
      }

      return (
        <FormProvider {...useFormMethods}>
          <form
            className={styles.form}
            ref={ref}
            onSubmit={useFormMethods.handleSubmit<F>(handleFormSubmit, handleFormError)}
          >
            {children(useFormMethods)}
          </form>
        </FormProvider>
      )
    })
  )

  ForwardRefWrapper.displayName = 'Form'

  return <ForwardRefWrapper {...props} />
}

Form.Item = FormItem
Form.Submit = FormSubmit

export default Form
