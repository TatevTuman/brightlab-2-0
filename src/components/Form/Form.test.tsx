import React from 'react'
import { render, cleanup, waitFor, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Button, Input } from '@elements'
import { emailPattern } from '@utils'
import Form from './Form'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Form', () => {
  const props = {
    onSubmit: jest.fn(),
    onError: jest.fn()
  }

  const form = () =>
    render(
      <Form<{ email: string }> {...props}>
        {useFormMethods => {
          return (
            <>
              <Input
                {...useFormMethods}
                type={'email'}
                name={'email'}
                label={'Email'}
                validation={{
                  required: { value: true, message: 'Email is required' },
                  pattern: { value: emailPattern, message: 'It doesn`t seems to be an email' }
                }}
              />
              <Button type={'secondary'} size={'md'} submit centered>
                Submit
              </Button>
            </>
          )
        }}
      </Form>
    )

  it('renders correctly', async () => {
    const { container } = form()
    const awaitedContainer = await waitFor(() => container)

    expect(awaitedContainer).toMatchSnapshot()
  })

  it('submit on invalid fields', async () => {
    const { getByText, getByLabelText } = form()

    await act(async () => {
      const email = await waitFor(() => getByLabelText('Email'))
      userEvent.type(email, 'invalid email')
      const submit = await waitFor(() => getByText('Submit'))
      userEvent.click(submit)
    })

    expect(props.onSubmit).toHaveBeenCalledTimes(0)
  })

  it('submit on valid fields', async () => {
    const { getByText, getByLabelText } = form()

    await act(async () => {
      const email = await waitFor(() => getByLabelText('Email'))
      userEvent.type(email, 'test@test.com')
      const submit = await waitFor(() => getByText('Submit'))
      userEvent.click(submit)
    })

    expect(props.onSubmit).toHaveBeenCalledTimes(1)
  })

  // TODO with different types of elements
})
