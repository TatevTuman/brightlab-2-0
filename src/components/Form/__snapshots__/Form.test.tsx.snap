// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="input"
    >
      <label
        data-required="true"
        for="email"
      >
        Email
      </label>
      <div
        class="inputInner"
      >
        <input
          autocomplete="on"
          data-cursor="true"
          data-prefix="false"
          data-suffix="false"
          id="email"
          name="email"
          placeholder="Type a text"
          tabindex="0"
          type="email"
          value=""
        />
        <div
          class="inputInnerPrefix"
          data-prefix="false"
        />
        <div
          class="inputInnerSuffix"
          data-suffix="false"
        />
      </div>
    </div>
    <button
      class="button button__secondary-md "
      data-centered="true"
      data-disabled="false"
      tabindex="0"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
